model/task/task.go

```diff
@@ -2514,13 +2515,22 @@ func (t *Task) Insert() error {
 // into the old_tasks collection. If this is a display task, its execution tasks
 // are also archived.
 func (t *Task) Archive() error {
+	if !utility.StringSliceContains(evergreen.TaskCompletedStatuses, t.Status) {
+		grip.Debug(message.Fields{
+			"message":   "task is in incomplete state, skipping archiving",
+			"task_id":   t.Id,
+			"execution": t.Execution,
+			"func":      "Archive",
+		})
+		return nil
+	}
 	if t.DisplayOnly && len(t.ExecutionTasks) > 0 {
 		return errors.Wrapf(ArchiveMany([]Task{*t}), "archiving display task '%s'", t.Id)
 	} else {
 		// Archiving a single task.
 		archiveTask := t.makeArchivedTask()
-		err := db.Insert(OldCollection, archiveTask)
-		if err != nil && !db.IsDuplicateKey(err) {
+		_, err := db.Upsert(OldCollection, bson.M{IdKey: MakeOldID(t.Id, t.Execution)}, archiveTask)
```

Comment:
Is the idea that we won't need this to be upsert after the deploy, or we will always want this? If the task was already archived for this execution then we wouldn't want to update right?

Reply:
I was thinking of keeping it since I reasoned that a dup key error here will block the possibility of completing the reset for that execution forever, and it seems that it [happens occasionally in prod currently](https://mongodb.splunkcloud.com/en-US/app/search/search?q=search%20index%3D%22evergreen%22%20old_tasks%20dup&display.page.search.mode=verbose&dispatch.sample_ratio=1&workload_pool=standard_perf&earliest=-7d%40h&latest=now&sid=1662673545.1073158).  But this seems like a bug specific to the `task-execution-timeout` job possibly picking up the same execution twice, which I don't think is a big problem as long as it succeeds one time.  I'll create another ticket to investigate / address this because I actually don't think it's related to this ticket change.

I think you're right - now that L2518 makes us return early if the task is unfinished, we don't run the risk of archiving an unfinished task and then getting stuck at the archive `Insert` step on the next attempt.  So it should be fine change this back to `Insert`

Reply:
For the task exec timeout job bug: https://jira.mongodb.org/browse/EVG-17676